CRM/Contribute/BAO/Contribution.php

```diff
@@ -533,12 +534,25 @@ public static function create(&$params) {
         'id' => $existingActivity['id'] ?? NULL,
       ], $campaignParams);
       if (!$activityParams['id']) {
-        // Don't set target contacts on update as these will have been
-        // correctly created and we risk overwriting them with
-        // 'best guess' params.
         $activityParams['source_contact_id'] = (int) ($params['source_contact_id'] ?? (CRM_Core_Session::getLoggedInContactID() ?: $contribution->contact_id));
         $activityParams['target_contact_id'] = ($activityParams['source_contact_id'] === (int) $contribution->contact_id) ? [] : [$contribution->contact_id];
       }
+      else {
+        // Check if target contact exists
+        $targetContactQuery = ActivityContact::get(FALSE)->setLimit(1)->setWhere([
```

Comment:
So this is obviously a high-volume function & we need to be mindful of queries. In particular unnecessary updates can be locking queries - so I would do a bit more work here to check whether we are actually making a change of the contribution->contact is already the target - probably by opening up ```['record_type_id:name', '=', 'Activity Targets'],``` to

```['record_type_id:name', 'IN', ['Activity Targets', 'Activity Sources']],```

and doing extra checks in the php layer. If it starts to get a bit long then extract it